packages/splitter/src/useSplitter.ts

```diff
@@ -0,0 +1,274 @@
+  });
+
+  const onTouchEnd = composeEventHandlers(props.onTouchEnd, () => {
+    // must remove global events on transaction finish
+    environment.removeEventListener('touchend', onTouchEnd);
+    environment.removeEventListener('touchmove', onTouchMove);
+  });
+
+  const onMouseDown = composeEventHandlers(props.onMouseDown, (event: React.MouseEvent) => {
+    event.preventDefault();
+    if (type === SplitterType.FIXED) {
+      if (dimension > min) {
+        setDimensionWithCache(min);
+      }
+      if (dimension < max) {
+        setDimensionWithCache(max);
+      }
+    } else {
+      // Must register global events to track mouse move outside the container
+      environment.addEventListener('mouseup', onMouseUp);
+      environment.addEventListener('mousemove', onMouseMove);
+    }
+  });
+
+  const onTouchStart = composeEventHandlers(props.onTouchStart, () => {
+    if (type === SplitterType.FIXED) {
+      if (dimension > min) {
+        setDimensionWithCache(min);
+      }
+      if (dimension < max) {
+        setDimensionWithCache(max);
+      }
+    } else {
+      // Must register global events to track mouse move outside the container
+      environment.addEventListener('touchend', onTouchEnd);
+      environment.addEventListener('touchmove', onTouchMove);
+    }
+  });
+
+  const toggleMinMax = useCallback(() => {
+    if (dimensionPreCacheRef.current === min) {
+      setDimensionWithCache(max);
+    } else {
+      setDimensionWithCache(min);
+    }
+  }, [dimensionPreCacheRef, setDimensionWithCache, min, max]);
+
+  const onKeyDown = composeEventHandlers(props.onKeyDown, (event: React.KeyboardEvent) => {
+    if (orientation === SplitterOrientation.VERTICAL) {
+      switch (event.key) {
+        case 'ArrowLeft':
+          type === SplitterType.VARIABLE && addDelta(-50);
+          break;
+        case 'ArrowRight':
+          type === SplitterType.VARIABLE && addDelta(50);
+          break;
+        case 'Enter':
+          toggleMinMax();
+          break;
+      }
+    } else {
+      switch (event.key) {
+        case 'ArrowUp':
+          type === SplitterType.VARIABLE && addDelta(-50);
+          break;
+        case 'ArrowDown':
+          type === SplitterType.VARIABLE && addDelta(50);
+          break;
+        case 'Enter': {
+          toggleMinMax();
+          break;
+        }
+      }
+    }
+  });
+
+  const getSeparatorProps = ({ ...other } = {}) => ({
+    'data-garden-container-id': 'containers.splitter',
+    'data-garden-container-version': PACKAGE_VERSION,
+    role: 'separator',
+    ref: separatorRef,
+    onMouseDown,
+    onKeyDown,
+    onTouchStart,
+    'aria-label': label,
+    'aria-labelledby': labelledBy,
+    'aria-controls': controls,
+    'aria-valuenow': dimension,
+    'aria-valuemin': min,
+    'aria-valuemax': max,
+    'aria-orientation': orientation,
+    tabIndex: 0,
+    ...other
+  });
+
+  return {
+    getSeparatorProps
+  };
```

Comment:
I had the same thoughts around the return value and how consumers of the hook would impact how a component renders & re-renders. Do you think a `useMemo` would prevent renders if the literal object is memoized?

Reply:
After doing some research on my end and seeing how references are treated, it seems that the returned object from `useSplitter` does indeed change on each render and might merit a `useMemo`. I have a sandbox I will post soon.

Reply:
Here are the following sandboxes:
* [Without useMemo](https://codesandbox.io/s/without-usememo-jktus?file=/src/components/splitter/useSplitter.ts:8772-8826)
* [With useMemo](https://codesandbox.io/s/with-usememo-6lxiv?file=/src/components/splitter/useSplitter.ts:9285-9303)
I added a "Force Render" button inside `App.tsx` and added a `useEffect` that dumps out a console log when the `splitterReturn` reference changed. There is a difference when `useMemo` is used. 

I did not include the event handlers in the `useMemo` for expediency purposes as they all need to be wrapped in `useCallback` to prevent un-necessary deps invalidation. In anycase, the splitter still works. Let me know both your thoughts @hzhu and @Francois-Esquire.

Reply:
@Francois-Esquire, my understanding is that when React state changes a component should re-render. And it's not necessary a bad thing. When the return value is memoized (in @mtomcal 's sandbox demos), the hook will return a new reference object each time `aria-valuenow` is updated (each time the splitter is moved). This would still cause a render (expected) and no renders are saved.

@mtomcal the execution of the hook seems expected since a component should re-render when (React) state changes. Memoizing the `useSplitter`'s return value in this scenario saves a returning value but the hook is still ran, its memory still allocated. Memoization in this scenario only saves "returning the value". Also, there could be an argument that an invocation to `useMemo` is more expensive than "a function returning a value".